sphinx_gallery/recommender.py

```diff
@@ -0,0 +1,308 @@
+        X : sparse matrix
+            A sparse matrix in CSR format of shape (n_samples, n_features) where
+            n_samples is the number of samples in the dataset and n_features is the
+            total number of features across all samples.
+        """
+        feature_names = []
+        all_values = defaultdict(list)
+        for row in data:
+            for feature_name, feature_value in row.items():
+                feature_names.append(feature_name)
+                all_values[feature_name].append(feature_value)
+
+        feature_names = sorted(set(feature_names))
+        data, indices, indptr = [], [], [0]
+        for row in data:
+            for j, feature_name in enumerate(feature_names):
+                if feature_name in row:
+                    feature_value = row[feature_name]
+                    data.append(feature_value)
+                    indices.append(j)
+            indptr.append(len(indices))
+        X = sparse.csr_matrix(
+            (data, indices, indptr), shape=(len(indptr) - 1, len(feature_names))
+        )
+        return X
+
+    @staticmethod
+    def tfidf_transformer(X):
+        """Transform a term frequency matrix into a term frequency-inverse document
+        frequency (TF-IDF) matrix.
+
+        Parameters
+        ----------
+        X : {ndarray, sparse matrix} of shape (n_samples, n_features)
+            A term frequency matrix.
+
+        Returns
+        -------
+        X_tfidf : {ndarray, sparse matrix} of shape (n_samples, n_features)
+            A tf-idf matrix of the same shape as X.
+        """
+        if not sparse.issparse(X):
+            X = sparse.csr_matrix(X, dtype=X.dtype)
+
+        n_samples, n_features = X.shape
+
+        # Count the number of non-zero values for each feature in sparse X
+        if sparse.isspmatrix_csr(X):
+            df = np.bincount(X.indices, minlength=n_features)
+        else:
+            df = np.diff(X.indptr)
+        df = df.astype(X.dtype, copy=False)
+        # perform idf smoothing
+        df += 1
+        n_samples += 1
+        idf = np.log(n_samples / df) + 1
+
+        idf_diag = sparse.diags(
+            idf,
+            offsets=0,
+            shape=(n_features, n_features),
+            format="csr",
+            dtype=X.dtype,
+        )
+        X_tfidf = X * idf_diag
+        X_tfidf = (X_tfidf.T / norm(X_tfidf, axis=1)).T
+        X_tfidf = sparse.csr_matrix(X_tfidf, dtype=X.dtype)
+
+        return X_tfidf
+
+    @staticmethod
+    def cosine_similarity(X, Y=None, dense_output=True):
+        """
+        Compute the cosine similarity between two vectors X and Y.
+
+        Parameters
+        ----------
+        X : {ndarray, sparse matrix} of shape (n_samples_X, n_features)
+            Input data.
+
+        Y : {ndarray, sparse matrix} of shape (n_samples_Y, n_features), default=None
+            Input data. If `None`, the output will be the pairwise
+            similarities between all samples in `X`.
```

Comment:
I would specialized `X` and `Y` to be a single type dense or sparse.
For the moment, the `tfidf_transformer` function will always return sparse matrix. So I would be inclined to only accept and expect sparse matrix.